src/name/ip_address.rs

```diff
@@ -14,6 +14,236 @@
 
 use crate::Error;
 
+#[cfg(feature = "alloc")]
+use alloc::string::String;
+
+const VALID_IP_BY_CONSTRUCTION: &str = "IP address is a valid string by construction";
+
+/// Either a Ipv4 or Ipv6 address, plus its owned string representation
+#[cfg(feature = "alloc")]
+#[derive(Clone, Debug, Eq, PartialEq, Hash)]
+pub enum IpAddress {
+    /// An ipv4 address and its owned string representation
+    IpV4Address(String, [u8; 4]),
+    /// An ipv6 address and its owned string representation
+    IpV6Address(String, [u8; 16]),
+}
+
+#[cfg(feature = "alloc")]
+impl AsRef<str> for IpAddress {
+    fn as_ref(&self) -> &str {
+        match self {
+            IpAddress::IpV4Address(ip_address, _) | IpAddress::IpV6Address(ip_address, _) => {
+                ip_address.as_str()
+            }
+        }
+    }
+}
+
+/// Either a ipv4 or ipv6 address, plus its borrowed string representation
+#[derive(Debug, Clone, Copy)]
+pub enum IpAddressRef<'a> {
+    /// An ipv4 address and its borrowed string representation
+    IpV4AddressRef(&'a [u8], [u8; 4]),
+    /// An ipv6 address and its borrowed string representation
+    IpV6AddressRef(&'a [u8], [u8; 16]),
+}
+
+#[cfg(feature = "alloc")]
+impl<'a> From<IpAddressRef<'a>> for IpAddress {
+    fn from(ip_address: IpAddressRef<'a>) -> IpAddress {
+        match ip_address {
+            IpAddressRef::IpV4AddressRef(ip_address, ip_address_octets) => IpAddress::IpV4Address(
+                String::from_utf8(ip_address.to_vec()).expect(VALID_IP_BY_CONSTRUCTION),
+                ip_address_octets,
+            ),
+            IpAddressRef::IpV6AddressRef(ip_address, ip_address_octets) => IpAddress::IpV6Address(
+                String::from_utf8(ip_address.to_vec()).expect(VALID_IP_BY_CONSTRUCTION),
+                ip_address_octets,
+            ),
+        }
+    }
+}
+
+// Returns the octets that correspond to the provided IPv4 address.
+//
+// This function can only be called on IPv4 addresses that have
```

Comment:
It would be nice to enforce this invariant in the type system by having a newtype wrapper.

Reply:
Given it is `pub(crate)` API, I didn't want to introduce many other types --impl specific and no external consumer can access this API-- but I agree the type system is the best place to enforce this. However, there's no way to avoid the usage of the inner type directly unless I'm missing something.

Reply:
Addressed in https://github.com/rustls/webpki/pull/21 (https://github.com/rustls/webpki/pull/21/commits/f1b6d66ba1fa8ef11aaf274ec24229280dcb17d9)